keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,987 @@
+-->
+
+<<[UNRESOLVED]>>
+
+The feature will be guarded by a feature gate and disabled by default in alpha. When alpha, surely 
+the Pod API field needs to be disabled, but I'm open to hear feedback on whether to disable scheduler's
+ComponentConfig API as well as the extension point logic.
+
+If we disable it, then users cannot use the new extension point at all.
+If we don't disable it, people can still customize their enqueue behaviors, like wire it with a
+customized field (like an annotation key).
+
+<<[/UNRESOLVED]>>
+
+<<[UNRESOLVED]>>
+
+As we don't write unschedulable condition to the Pod's spec if it carries `.spec.schedulePaused=true`,
+it may provider better UX to add a "paused" column to the output of `kubectl get pod (-o wide)`.
+
+<<[/UNRESOLVED]>>
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+End-users may be confused by no scheduling events in the output of `kubectl describe pod xyz`. We
+should provide detailed documentation, along with metrics, tooling (kubectl) and(or) events.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### API
+
+A new API field `schedulePaused` will be added to Pod's spec:
+
+```go
+type PodSpec struct {
+    SchedulePaused *bool // <-- New field.
+}
+```
+
+In the scheduler's ComponentConfig API, we'll add a new type of extension point called `Enqueue`:
+
+```go
+type Plugins struct {
+    PreFilter PluginSet
+    ......
+    Enqueue PluginSet // <-- New field.
+}
+```
+
+### Implementation
+
+Inside scheduler, an internal queue called "activeQ" is designed to store all ready-to-schedule Pods.
+In this KEP, we're going to wire the aforementioned `Enqueue` extension to the logic deciding whether
+or not to add Pods to "activeQ".
+
+Specifically, prior to adding a Pod to "activeQ", scheduler iterates over registered Enqueue plugins.
+An Enqueue plugin must implement the `EnqueuePlugin` interface to return a `Status` to tell scheduler
+whether this Pod can be admitted or not:
+
+```go
+// EnqueuePlugin is an interface that must be implemented by "EnqueuePlugin" plugins.
+// These plugins are called prior to adding Pods to activeQ.
+type EnqueuePlugin interface {
+    Plugin
+    PreEnqueue(ctx context.Context, state *CycleState, p *v1.Pod) *Status
+}
+```
+
+A Pod can be moved to activeQ only when all Enqueue plugins returns `Success`. Otherwise, it's moved
```

Comment:
Move the pods rejected by default enqueue plugin back to backoffQ/activeQ seems useless when the `spec.SchedulePaused` remains in true. We should add some filter logic at that moment.

Reply:
You mean the current logic of `flushUnschedulablePodsLeftover`? Yup, we should tweak the impl. logic a bit. But in terms of KEPing, it's way low-level impl., and also someday we may remove it, I suppose we shouldn't specify the details in this doc.